# <center>MC906/MO416 - Introduction to Artificial Intelligence</center>
# <center>Project 1 - Pacman with AI</center>
# <center>Institute of Computing - Unicamp</center>
# <center>Prof. Esther Colombini</center>

### <center>Eduardo S. Ito (RA159086)</center>
### <center>Lucas Peres  (RA 265193)</center>
### <center>Thales E. Nazatto (RA 074388)</center>

# Conteúdo

1. Introdução
    1. Descrição do problema
2. Metodologia
    1. Modelagem da solução
        1. Requisitos
        2. Implementação
        3. Problema, agentes e ambiente
    2. Algoritmos
        1. Busca não-informada
        2. Busca informada
        3. Busca local (*Hill Climbing*)
3. Resultados
4. Conclusão
5. Referências

# Introdução
**Pac-man** é um jogo para *Arcades* lançado em 1980 pela **Namco**, criado por Toru Iwatani e ilustrado na Figura 1. Nele, o jogador controla uma bolinha amarela (o *Pac-Man* do título) dentro de um labirinto que tem como objetivo comer todas as bolinhas brancas existentes nele enquanto é perseguido por 4 fantasmas de nomes *Blinky*, *Pinky*, *Inky* e *Clyde*. Na época, Iwatani criou o jogo para atrair pessoas de ambos os sexos, uma vez que os gêneros de jogos lançados nessa época, como jogos de guerra e esportes, eram majoritariamente pensados para agradar o público masculino. Tal estratégia se provou um grande sucesso, e *Pac-Man* se tornou uma das mais lucrativas franquias de todos os tempos, gerando mais de US$ 14 bilhões em receita até o ano de 2016 **[1]**.

<img src="https://thumbs.web.sapo.io/?W=775&H=0&delay_optim=1&webp=1&epic=NGRiQ8y+7rDb2cocSbMwl47ynqxTFYLIDm2IhR5IZK5BFQooKY19ghBWe//2n+M5C2MgCiqhexDuDLBr/f/MrofnNA==" height="240" width="320" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 1.</strong> Tela do jogo <i>Pac-Man</i></center>

Uma dos grandes inovações para a época foi sua Inteligência Artificial. Cada fantasma se comportava de maneira diferente de acordo com os movimentos realizados pelo jogador, que poderiam ter a possibilidade de encurralá-lo. Enquanto *Blinky* perseguia *Pac-Man* diretamente, *Pinky* e *Inky* tentavam se posicionar a sua frente e *Clyde* alternava seus estados entre perseguir o *Pac-Man* e fugir dele. Tais comportamentos fizeram com que *Pac-Man* fosse um caso de estudos no ramo da Inteligência Artificial.



Neste relatório, será mostrada uma modelagem do problema determinado em **[2]** utilizando 5 algoritmos de busca no contexto deste jogo, sendo 2 de busca não-informada (*Breadth-First Search* e *Depth-First Search*), 2 de busca informada (*Greedy Best-First Search* e *A*\*) e 1 de busca local (*Hill Climbing*). Primeiro será detalhada como foi feita a modelagem do problema e as soluções aplicadas. Depois, serão feitos experimentos e dicussões para avaliar o comportamento dos algoritmos e, no final, as conclusões destes experimentos.

## Descrição do problema

Conforme determinado em **[2]**, o problema consiste em implementar 5 algoritmos diferentes de busca, sendo 2 de busca não-informada, 2 de busca informada (com 2 heurísticas diferentes) e 1 de busca local usando o *Pac-man* como tema, tendo como objetivo comer todas as bolinhas presentes no labirinto, caso consiga evitar os fantasmas. Dentre esses, é necessário avaliar critérios como otimalidade, completude e custo computacional para definir qual a melhor solução dentre elas.

O labirinto é representado por um grid, também ilustrado na Figura 2, cujo tamanho e forma devem ser definidos. Deve ser definido também, além das posições das bolinhas que o *Pac-man* deve comer durante a execução do algoritmo, as posições de 3 fantasmas (que serão estáticos) e uma posição final de parada. É necessário especificar também a representação dos estados, as ações realizadas, o teste realizado para se chegar ao objetivo, o custo **g(x)** do caminho percorrido e as heurísticas utilizadas.

<img src="https://shaunlebron.github.io/pacman-mazegen/img/origmaps_2x_print.png" height="360" width="480" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 2.</strong> Exemplos de labirintos semelhantes <strong>[6]</strong></center>

# Metodologia

Nele, a classe **SimplePacmanProblem** é uma classe filha da classe **Problem**, que é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, obstáculos e heurística. **action_cost** é a função heurística utilizada para algoritmos de busca informada (ou **g(x)**), **h** é a função heurística utilizada para o *A*\* (ou **h(x)**) e **value** é a função heurística utilizada para algoritmos de busca locais (ou **g(x)**). No AIMA, o heurística nos algoritmos de busca locais em seu código original é realizada em direção a um mínimo local. O valor de **directions** seria semelhante às ações que o *Pac-man* executaria no caminho, fazendo estados e ações serem tuplas no formato **(x,y)** e praticamente não tendo diferenciação.

A classe semelhante no código utilizado neste projeto é a **PositionSearchProblem2**, que é uma classe filha da classe **SearchProblem2**, a qual é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, heurística e o estado do jogo, que guarda as posições das paredes, fantasmas, das bolinhas e do *Pac-man*. **getCostOfActions** é a função utilizada para determinar os estados através da função heurística para os algoritmos de busca informada e local (dependendo dos parâmetros utilizados pode ser **g(x)** ou **h(x)**). **getSuccessors** é a função que disponibiliza os nós sucessores da busca, semelhante a função **actions** presente na AIMA. Há ainda a função **isGoalState**, semelhante a função **goal_test** da classe **Problem**. A classe **Search2Agent** faz o papel de agente inteligente e a ligação entre problema, algoritmo de busca, estado do jogo e heurística.

Um diagrama de classes com todas as classes presentes no código pode ser visto no PDF abaixo:


In [1]:
# This .pdf file shows class diagram that implements how the problem was modeled using game graphical interface. It is part
# of how requirement [R4] and requirement [8] was modeling from Berkeley algorithm design base, and modification done
# to comply with MO416 Project 1 requirements.
import os
cwd = os.getcwd()
cwd=cwd+'/modelling/modellingClasses.pdf'
print(cwd)
import webbrowser
webbrowser.open(r''+os.path.join('modelling','modellingClasses.pdf'))
webbrowser.open(cwd)

C:\Users\Thales E. Nazatto\Documents\Pós\MO416A\MO416AI\Project1/modelling/modellingClasses.pdf


True

## Algoritmos

## Precondição: bibliotecas numpy e tkinter instaladas

### Busca não-informada

#### Depth-First Search

In [32]:
# [R1.1] Show dfs (depth first search) method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=dfs -z .6

[Search2Agent] using function dfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 72 in 0.009758710861206055s
[R13] Search nodes expanded: 79
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (3, 29), (2, 29), (1, 29), (1, 28), (1, 27), (1, 26), (1, 25), (1, 24), (1, 23), (1, 22), (1, 21), (1, 20), (2, 20), (3, 20), (4, 20), (5, 20), (6, 20), (6, 19), (6, 18), (6, 17), (6, 16), (5, 16), (4, 16), (3, 16), (2, 16), (1, 16), (1, 17), (1, 18), (1, 19), (3, 15), (3, 14), (3, 13), (2, 13), (1, 13), (1, 12), (1, 11), (1, 10), (2, 10), (3, 10), (3, 9), (3, 8), (3, 7), (2, 7), (1, 7), (4, 7), (5,

#### Breadth-First Search

In [33]:
# [R1.2] Show bfs (breadth first search) method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=bfs -z .5

[Search2Agent] using function bfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 48 in 0.04782390594482422s
[R13] Search nodes expanded: 336
[R13] Nodes visited: [(23, 29), (23, 28), (24, 29), (22, 29), (23, 27), (25, 29), (21, 29), (23, 26), (26, 29), (20, 29), (22, 26), (26, 28), (19, 29), (21, 26), (26, 27), (18, 29), (21, 25), (26, 26), (17, 29), (21, 24), (26, 25), (16, 29), (21, 23), (26, 24), (15, 29), (22, 23), (20, 23), (26, 23), (15, 28), (14, 29), (23, 23), (19, 23), (26, 22), (15, 27), (13, 29), (23, 22), (18, 23), (26, 21), (15, 26), (12, 29), (23, 21), (18, 24), (18, 22), (17, 23), (26, 20), (16, 26), (12, 28), (11, 29), (23, 20), (18, 25), (18, 21), (16, 23), (26, 19), (25, 20), (17, 26), (12, 27), (10, 29), (24, 20), (22, 20), (18, 2

### Busca informada

#### A*

In [34]:
# [R2.1] Show A* and Manhattan heuristic method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .5

[Search2Agent] using function astar and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 48 in 0.040992021560668945s
[R13] Search nodes expanded: 202
[R13] Nodes visited: [(23, 29), (23, 28), (22, 29), (23, 27), (21, 29), (23, 26), (20, 29), (22, 26), (19, 29), (21, 26), (18, 29), (21, 25), (17, 29), (21, 24), (16, 29), (21, 23), (15, 29), (20, 23), (15, 28), (14, 29), (19, 23), (15, 27), (13, 29), (18, 23), (15, 26), (12, 29), (18, 22), (17, 23), (12, 28), (11, 29), (18, 21), (16, 23), (12, 27), (10, 29), (18, 20), (15, 23), (12, 26), (9, 29), (18, 19), (14, 23), (11, 26), (8, 29), (18, 18), (17, 19), (13, 23), (10, 26), (7, 29), (18, 17), (16, 19), (12, 23), (9, 26), (6, 29), (18, 16), (15, 19), (11, 23), (9, 25), 

In [19]:
# [R2.1] Show A* and Euclidean heuristic method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6

[Search2Agent] using function astar and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 48 in 0.053681373596191406s
[R13] Search nodes expanded: 263
[R13] Nodes visited: [(23, 29), (23, 28), (22, 29), (23, 27), (21, 29), (23, 26), (22, 26), (20, 29), (21, 26), (19, 29), (24, 29), (21, 25), (21, 24), (18, 29), (21, 23), (17, 29), (20, 23), (19, 23), (16, 29), (18, 23), (25, 29), (18, 22), (15, 29), (17, 23), (15, 28), (18, 21), (15, 27), (15, 26), (22, 23), (16, 23), (18, 20), (14, 29), (18, 19), (15, 23), (17, 19), (13, 29), (18, 18), (14, 23), (16, 19), (18, 24), (18, 17), (26, 29), (12, 29), (15, 19), (12, 28), (13, 23), (26, 28), (12, 27), (18, 16), (16, 26), (12, 26), (14, 19), (26, 27), (23, 23), (18, 15), (12,

#### Greedy Best-First Search

In [20]:
#[R2.2] Show gbfs (greedy best first search) and Manhattan heuristic method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6

[Search2Agent] using function gbfs and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 52 in 0.00780797004699707s
[R13] Search nodes expanded: 61
[R13] Nodes visited: [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (18, 22), (18, 21), (18, 20), (18, 19), (18, 18), (18, 17), (18, 16), (18, 15), (18, 14), (18, 13), (18, 12), (18, 11), (18, 10), (18, 9), (18, 8), (18, 7), (18, 6), (18, 5), (18, 4), (17, 4), (16, 4), (15, 4), (15, 3), (14, 4), (13, 4), (12, 4), (12, 3), (11, 4), (10, 4), (9, 4), (9, 5), (9, 6), (12, 2), (9, 7), (12, 1), (11, 1), (10, 1), (9, 1), (8, 1), (7, 1), (6, 1), (6, 2), (6, 3), (5, 1), (4, 1), (3, 1), (2, 1), (1, 1), (1, 2),

In [35]:
##[R2.2] Show gbfs (greedy best first search) and Euclidean heuristic method.
!python pacman.py -l defaultLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .5

[Search2Agent] using function gbfs and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? True
[R16] Path found with total cost g(x) of 48 in 0.010736942291259766s
[R13] Search nodes expanded: 72
[R13] Nodes visited: [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (18, 22), (18, 21), (18, 20), (18, 19), (17, 19), (16, 19), (15, 19), (14, 19), (14, 18), (14, 17), (14, 16), (14, 15), (13, 15), (12, 15), (11, 15), (11, 16), (12, 16), (11, 17), (13, 16), (12, 17), (13, 17), (15, 15), (15, 16), (13, 18), (16, 15), (15, 17), (16, 16), (13, 19), (12, 19), (11, 19), (10, 19), (9, 19), (9, 18), (9, 17), (9, 16), (9, 15), (9, 14), (9, 13), (9, 12), (9, 11), (9, 10), (8, 10), (7, 10), (6, 10), (6, 9), (

### Busca local (*Hill Climbing*)

In [118]:
from re import search
from subprocess import run

def run_hcs(layout):
    '''
    Extrai o número de nós visitados e expandidos da
    stdout do comando pacman.py.
    '''
    nodes_visited = 0
    nodes_expanded = 0
    output = run(['python', 'pacman.py', 
                  '-l', layout, 
                  '-p', 'Search2Agent', 
                  '-a', 'fn=hcs', 
                  '-z', '.6'], capture_output=True)
    results = output.stdout.decode('utf-8').split('\n')
    for result in results:
        matchstates = search('Solution states: ', result)
        if matchstates:
            nodes_visited = int(result[matchstates.end():].split(' ')[0])
        matchnodes = search('nodes expanded: ', result)
        if matchnodes:
            nodes_expanded = int(result[matchnodes.end():])
    return nodes_visited, nodes_expanded

def mean_hcs(n, layout):
    '''
    Executa Hill Climbing n vezes no layout escolhido
    e retorna a média de nós visitados e expandidos
    nas n execuções.
    '''
    totalvisited = 0
    totalexpanded = 0
    for _ in range(n):
        nvisited, nexpanded = run_hcs('easyLayout')
        totalvisited += nvisited
        totalexpanded += nexpanded
    meanvisited = totalvisited / n
    meanexpanded = totalexpanded / n
    print(f'{layout}\tNo. nodes visited (mean): {meanvisited}')
    print(f'{layout}t\tNo. nodes expanded (mean): {meanexpanded}')
    return meanvisited, meanexpanded

In [119]:
nvisited, nexpanded = mean_hcs(2, 'easyLayout')

NODES VISITED 55
NODES VISITED 73
easyLayout	No. nodes visited (mean): 64.0
easyLayoutt	No. nodes expanded (mean): 148.0


# Resultados

### *defaultLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|Complexidade|
|---|---|---|---|---|---|
|Depth-first Search                  |Não|Sim|79 |73|$O(.)$|
|Breadth-first Search                |Não|Não|336|49|$O(.)$|
|A* (Euclidean)                      |Sim|Sim|263|49|$O(.)$|
|Greedy Best-first Search (Euclidean)|Não|Sim|72 |49|$O(.)$|
|A* (Manhattan)                      |Sim|Sim|202|49|$O(.)$|
|Greedy Best-first Search (Manhattan)|Não|Sim|61 |53|$O(.)$|
|Hill Climbing                       |Não|Sim| ? |? |$O(.)$|


### *easyLayout*

### *moderateLayout*

### *hardLayout*
Apenas Himm Climbing consegue, e por sorte.

# Vídeo

In [40]:
import os
def play_movie(path):
    from os import startfile
    startfile(path)
    
class Video(object):
    def __init__(self,path):
        self.path = path

    def play(self):
        from os import startfile
        startfile(self.path)

class Movie_MP4(Video):
    type = "MP4"

path =  os.getcwd()
path = path+"\\video\mo416Project1video.mp4"
movie = Movie_MP4(path)
movie.play()

# Note: Not all use cases were displayed in video as there is a limitation of 25MB storage in github.

In [38]:
# Bibliografia

**[1]** Pac-man - Wikipedia. URL: https://en.wikipedia.org/wiki/Pac-Man

**[2]** Project 1. URL: https://drive.google.com/file/d/187lgVekPC0kBmA2AjOm-KkV8Q_lFLeoH/view

**[3]** Search for AIMA 4th edition - Implementation of search algorithms and search problems for AIMA. URLs: https://github.com/aimacode/aima-python/blob/master/search4e.ipynb, https://github.com/aimacode/aima-python/blob/master/search.ipynb

**[4]** UC Berkeley CS188 Intro to AI -- Course Materials. URL: http://ai.berkeley.edu/search.html

**[5]** *tkint* documentation. URL: https://docs.python.org/3/library/tkinter.html

**[6]** Pac-Man Maze Generation. URL: https://shaunlebron.github.io/pacman-mazegen

SyntaxError: invalid syntax (<ipython-input-38-aa7878771861>, line 3)